File: apps/tag/web-vite/src/hooks/useFormValidation.ts
import { useState, useCallback } from 'react'

export interface ValidationRule {
  required?: boolean
  minLength?: number
  maxLength?: number
  pattern?: RegExp
  custom?: (value: any) => string | null
}

export interface ValidationRules {
  [key: string]: ValidationRule
}

export interface ValidationErrors {
  [key: string]: string
}

export function useFormValidation<T extends Record<string, any>>(
  initialValues: T,
  validationRules: ValidationRules
) {
  const [values, setValues] = useState<T>(initialValues)
  const [errors, setErrors] = useState<ValidationErrors>({})
  const [touched, setTouched] = useState<Record<string, boolean>>({})
  const [isSubmitting, setIsSubmitting] = useState(false)

  const validateField = useCallback((name: string, value: any): string | null => {
    const rule = validationRules[name]
    if (!rule) return null

    // Required validation
    if (rule.required && (!value || value.trim() === '')) {
      return 'Este campo é obrigatório'
    }

    // Min length validation
    if (rule.minLength && value && value.length < rule.minLength) {
      return `Mínimo de ${rule.minLength} caracteres`
    }

    // Max length validation
    if (rule.maxLength && value && value.length > rule.maxLength) {
      return `Máximo de ${rule.maxLength} caracteres`
    }

    // Pattern validation
    if (rule.pattern && value && !rule.pattern.test(value)) {
      return 'Formato inválido'
    }

    // Custom validation
    if (rule.custom) {
      return rule.custom(value)
    }

    return null
  }, [validationRules])

  const validateForm = useCallback((): boolean => {
    const newErrors: ValidationErrors = {}
    let isValid = true

    Object.keys(validationRules).forEach(fieldName => {
      const error = validateField(fieldName, values[fieldName])
      if (error) {
        newErrors[fieldName] = error
        isValid = false
      }
    })

    setErrors(newErrors)
    return isValid
  }, [values, validationRules, validateField])

  const handleChange = useCallback((name: string, value: any) => {
    setValues(prev => ({ ...prev, [name]: value }))
    
    // Clear error when user starts typing
    if (errors[name]) {
      setErrors(prev => ({ ...prev, [name]: '' }))
    }
  }, [errors])

  const handleBlur = useCallback((name: string) => {
    setTouched(prev => ({ ...prev, [name]: true }))
    
    // Validate field on blur
    const error = validateField(name, values[name])
    setErrors(prev => ({ ...prev, [name]: error || '' }))
  }, [values, validateField])

  const handleSubmit = useCallback(async (onSubmit: (values: T) => Promise<void>) => {
    setIsSubmitting(true)
    
    try {
      if (validateForm()) {
        await onSubmit(values)
      }
    } catch (error) {
      console.error('Form submission error:', error)
    } finally {
      setIsSubmitting(false)
    }
  }, [values, validateForm])

  const resetForm = useCallback(() => {
    setValues(initialValues)
    setErrors({})
    setTouched({})
    setIsSubmitting(false)
  }, [initialValues])

  const clearErrors = useCallback(() => {
    setErrors({})
  }, [])

  const getFieldProps = useCallback((name: string) => ({
    value: values[name],
    onChange: (value: any) => handleChange(name, value),
    onBlur: () => handleBlur(name),
    error: errors[name],
    isTouched: touched[name],
    hasError: touched[name] && errors[name],
  }), [values, errors, touched, handleChange, handleBlur])

  return {
    values,
    errors,
    touched,
    isSubmitting,
    handleChange,
    handleBlur,
    handleSubmit,
    resetForm,
    getFieldProps,
    validateForm,
    clearErrors,
  }
}

File: apps/tag/flutter-v1.5.0/lib/hooks/use_form_validation.dart
import 'package:flutter/material.dart';

class ValidationRule {
  final bool required;
  final int? minLength;
  final int? maxLength;
  final RegExp? pattern;
  final String? Function(String?)? custom;

  const ValidationRule({
    this.required = false,
    this.minLength,
    this.maxLength,
    this.pattern,
    this.custom,
  });
}

class FormValidationHook {
  final Map<String, TextEditingController> controllers;
  final Map<String, String?> errors;
  final Map<String, bool> touched;
  final bool isSubmitting;
  final Function(String, String?) setError;
  final Function(String) setTouched;
  final Function(bool) setSubmitting;

  FormValidationHook({
    required this.controllers,
    required this.errors,
    required this.touched,
    required this.isSubmitting,
    required this.setError,
    required this.setTouched,
    required this.setSubmitting,
  });

  String? validateField(String fieldName, String? value, ValidationRule rule) {
    // Required validation
    if (rule.required && (value == null || value.trim().isEmpty)) {
      return 'Este campo é obrigatório';
    }

    if (value != null && value.isNotEmpty) {
      // Min length validation
      if (rule.minLength != null && value.length < rule.minLength!) {
        return 'Mínimo de ${rule.minLength} caracteres';
      }

      // Max length validation
      if (rule.maxLength != null && value.length > rule.maxLength!) {
        return 'Máximo de ${rule.maxLength} caracteres';
      }

      // Pattern validation
      if (rule.pattern != null && !rule.pattern!.hasMatch(value)) {
        return 'Formato inválido';
      }

      // Custom validation
      if (rule.custom != null) {
        return rule.custom!(value);
      }
    }

    return null;
  }

  bool validateForm(Map<String, ValidationRule> rules) {
    bool isValid = true;
    
    for (String fieldName in rules.keys) {
      final controller = controllers[fieldName];
      if (controller != null) {
        final error = validateField(fieldName, controller.text, rules[fieldName]!);
        setError(fieldName, error);
        if (error != null) {
          isValid = false;
        }
      }
    }
    
    return isValid;
  }

  void handleFieldChange(String fieldName) {
    // Clear error when user starts typing
    if (errors[fieldName] != null) {
      setError(fieldName, null);
    }
  }

  void handleFieldBlur(String fieldName, ValidationRule rule) {
    setTouched(fieldName);
    
    // Validate field on blur
    final controller = controllers[fieldName];
    if (controller != null) {
      final error = validateField(fieldName, controller.text, rule);
      setError(fieldName, error);
    }
  }

  Future<bool> handleSubmit(
    Map<String, ValidationRule> rules,
    Future<void> Function(Map<String, String>) onSubmit,
  ) async {
    setSubmitting(true);
    
    try {
      if (validateForm(rules)) {
        final values = <String, String>{};
        for (String fieldName in controllers.keys) {
          values[fieldName] = controllers[fieldName]!.text;
        }
        
        await onSubmit(values);
        return true;
      }
      return false;
    } catch (error) {
      debugPrint('Form submission error: $error');
      return false;
    } finally {
      setSubmitting(false);
    }
  }

  void resetForm() {
    for (var controller in controllers.values) {
      controller.clear();
    }
    // Reset errors and touched states
    // This would need to be implemented in the parent widget
  }

  void clearErrors() {
    // Clear all errors
    // This would need to be implemented in the parent widget
  }
}
